precompiles/erc20/utils_test.go

```diff
@@ -110,17 +206,303 @@ func (s *PrecompileTestSuite) requireSendAuthz(grantee, granter sdk.AccAddress,
+//
+// In case of a direct call to the precompile, the precompile's ABI is used. Otherwise, the
+// ERC20CallerContract's ABI is used and the given contract address.
+func (is *IntegrationTestSuite) getTxAndCallArgs(
+	callType CallType,
+	contractData ContractsData,
+	methodName string,
+	args ...interface{},
+) (evmtypes.EvmTxArgs, factory.CallArgs) {
+	cd := contractData.GetContractData(callType)
+
+	txArgs := evmtypes.EvmTxArgs{
+		To: &cd.Address,
+	}
+
+	callArgs := factory.CallArgs{
+		ContractABI: cd.ABI,
+		MethodName:  methodName,
+		Args:        args,
+	}
+
+	return txArgs, callArgs
+}
+
+// ExpectedBalance is a helper struct to check the balances of accounts.
+type ExpectedBalance struct {
+	address  sdk.AccAddress
+	expCoins sdk.Coins
+}
+
+// ExpectBalances is a helper function to check if the balances of the given accounts are as expected.
+func (is *IntegrationTestSuite) ExpectBalances(expBalances []ExpectedBalance) {
+	for _, expBalance := range expBalances {
+		for _, expCoin := range expBalance.expCoins {
+			coinBalance, err := is.handler.GetBalance(expBalance.address, expCoin.Denom)
+			Expect(err).ToNot(HaveOccurred(), "expected no error getting balance")
+			Expect(coinBalance.Balance.Amount.Int64()).To(Equal(expCoin.Amount.Int64()), "expected different balance")
+		}
+	}
+}
+
+// ExpectBalancesForContract is a helper function to check expected balances for given accounts depending
+// on the call type.
+func (is *IntegrationTestSuite) ExpectBalancesForContract(callType CallType, contractData ContractsData, expBalances []ExpectedBalance) {
+	switch {
+	case slices.Contains(nativeCallTypes, callType):
+		is.ExpectBalances(expBalances)
+	case slices.Contains(erc20CallTypes, callType):
+		is.ExpectBalancesForERC20(callType, contractData, expBalances)
+	default:
+		panic("unknown contract call type")
+	}
+}
+
+// ExpectBalancesForERC20 is a helper function to check expected balances for given accounts
+// when using the ERC20 contract.
+func (is *IntegrationTestSuite) ExpectBalancesForERC20(callType CallType, contractData ContractsData, expBalances []ExpectedBalance) {
+	contractABI := contractData.GetContractData(callType).ABI
+
+	for _, expBalance := range expBalances {
+		for _, expCoin := range expBalance.expCoins {
+			addr := common.BytesToAddress(expBalance.address.Bytes())
+
+			txArgs, callArgs := is.getTxAndCallArgs(callType, contractData, "balanceOf", addr)
+
+			passCheck := testutil.LogCheckArgs{ExpPass: true}
+
+			_, ethRes, err := is.factory.CallContractAndCheckLogs(contractData.ownerPriv, txArgs, callArgs, passCheck)
+			Expect(err).ToNot(HaveOccurred(), "expected no error getting balance")
+
+			var balance *big.Int
+			err = contractABI.UnpackIntoInterface(&balance, "balanceOf", ethRes.Ret)
+			Expect(err).ToNot(HaveOccurred(), "expected no error unpacking balance")
+			Expect(balance.Int64()).To(Equal(expCoin.Amount.Int64()), "expected different balance")
```

Comment:
looks like only this needs to be inside the second `for`?

Reply:
the whole inner loop actually can be removed since the ERC20 contract only holds one balance for the token denom anyways 👍